### PR TITLE
fix(cache): clear-all freed nothing — the trim target was the one number PMS refuses

### PR DESCRIPTION
**Pull request by trinadhthatakula**
## The report

Clear-all-cache freed nothing and always answered **"there was no cache left to clear"**, on four devices with gigabytes of cache present: stock Android 16 and 17, HyperOS 3 (MIUI, Android 16), and Samsung Android 16 — some rooted, some on Shizuku. One root-cause chain explains all four, which is why the symptom was identical across privilege modes rather than varying by one.

## Root cause 1 — the target was the one number PMS is guaranteed to refuse

**`pm trim-caches N` means "ensure N bytes are *usable*", not "free N bytes of cache."** The first line of `FreeStorageHelper.freeStorage` is:

```java
if (file.getUsableSpace() >= bytes) return;
```

No reserve is added to that comparison. A target at or below current free space is a silent no-op — and `pm` still exits 0.

What v1.94 passed was `StorageStatsManager.getFreeBytes()`, which is `usableSpace + max(0, cacheTotal - cacheReserve)`. The cache reserve is a percentage of **total** storage — roughly 23 GB on a 232 GB phone — so it exceeds any device's entire app cache, the clearable term is `0`, and the target collapses to `usableSpace` **exactly**. PMS returned on its first line, every time, on every device.

This is now `usableSpace + totalCacheBytes()` — strictly greater than what PMS measures, by exactly the amount to reclaim.

Free space is read from `Environment.getDataDirectory().usableSpace` rather than another storage API deliberately: it has to be the *same* `File` that `StorageManager.findPathForUuid(UUID_PRIVATE_INTERNAL)` resolves to, since the whole defect was a target sitting a hair below what PMS measured. Lint's `UsableSpace` check asks for `getAllocatableBytes` instead — that advice is correct for its stated case (will this file fit?) and is precisely the mistake being undone here, since `getAllocatableBytes` adds the clearable cache back. Suppressed with that reasoning in place.

**Bounding the target is still load-bearing, just at the other end.** `freeStorage` is an escalating ladder and app cache is rungs 4 and 8; overshooting continues into pruning unused static shared libraries (rung 5) and uninstalling instant apps (rungs 7 and 10). A sum of the two real numbers stops at rung 8; `pm trim-caches 100G` does not. (Correcting a claim I made on #373: rungs 2 and 3 need `FLAG_ALLOCATE_AGGRESSIVE`, which `pm trim-caches` never sets, and rung 6 — dexopt output — is `// TODO: Implement` upstream. The collateral is smaller than I previously warned, but not zero.)

## Root cause 2 — root's working rung was unreachable

`pm trim-caches` exits 0 unconditionally: `PackageManagerShellCommand.runTrimCaches` waits on its observer and then `return 0`s with no regard for how many bytes moved. `RootSystemGateway.clearAllCaches` did:

```kotlin
val trim = runCommand("pm trim-caches $targetFreeBytes")
if (trim.isSuccess) return trim          // ← always taken
```

So the trim's meaningless 0 short-circuited the `rm -rf` sweep beneath it. Root reported success on a no-op and never ran the rung that works.

The verdict is now logged and dropped, and the sweep always runs. The two rungs were never a fallback chain — there is no success to fall back *from* when the exit code cannot fail. The trim still goes first because it is PMS doing the deleting (volumes Thor's globs do not name, every Android user, accounting left current); its failure is worth a log line and nothing more.

Same defect class as 2fded55 on #373 — trusting an unverified success signal — reintroduced two files away, one commit later.

## Shizuku

Not a dead end, and it needed no new mechanism: an emulator run as uid 2000 (shell, which is Shizuku's uid) freed the cache correctly once given a correct `N`. What it does need is the **cache total**, and only `PACKAGE_USAGE_STATS` can measure that — there is no permission-free way to size total app cache (`dumpsys diskstats` reports partition free space only).

The null-target message named the wrong missing thing (free space, which needs no permission). It now names Usage Access, which is the action the user can actually take. Guessing the other half of the sum is not an option: too low is a no-op, too high walks into the non-cache rungs.

Root survives a null target — its sweep names the directories itself. Shizuku is the one place the op is a hard requirement.

## Dhizuku

Unchanged. `clearAllCaches` never consults the target and still returns the reason it already gave.

## Tests

`cacheTrimTarget` is split out as a pure `internal` function for the same reason `RootFreezeChainTest` exists — `RootSystemGateway` and `StorageStatsHelper` are two framework calls from untestable, so the arithmetic that failed is tested on its own. `CacheTrimTargetTest` pins it at device-shaped magnitudes: the regression is unreproducible at toy scale, because it only appears once the cache reserve dwarfs the cache. One case reconstructs `getFreeBytes()`'s arithmetic on the reported device and asserts it lands at exactly free space, so a future collapse back to a single framework call has something concrete to trip over.

**855 tests green, lint clean, androidTest compiles.** No `versionCode` change.

## Scope

Per the reported behaviour, not a redesign. No per-app loop, no USER/SYSTEM filtering — the one-shot sweep is the right shape once it is reachable.

## Still owed elsewhere

The release-notes line retracting `release-notes/v1.94.0/github.md:84` at the next `master` merge (band A row 1) is unaffected by this and still outstanding.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved whole-volume cache clearing reliability, including cleanup after automatic trimming fails.
  * Corrected cleanup target calculations using available storage and measured cache usage.
  * Improved guidance when Usage Access is unavailable and clarified that command success does not guarantee cache removal.
  * Ensured accurate handling of empty and zero-cache scenarios.
* **Localization**
  * Added translated messages explaining cache-clearing requirements and unsupported operations.
* **Tests**
  * Added coverage for cleanup calculations, regression, and edge cases.

### CODE REVIEW

**Comment by trinadhthatakula** [comment]
## Self-review

Three findings. One fixed in `b2d5967`, one verified-and-left, one verified-and-closed.

### 1. Fixed — the trim target re-measured the cache the baseline had just measured

Introduced by this PR. Folding `totalCacheBytes()` inside `cacheTrimTargetBytes()` gave `clearAllCaches` **three** `queryStatsForUser` calls where two suffice — target, before, after — with the first two microseconds apart.

Two costs, and the second is the one that matters:

- `installd`'s `collectManualStats` is a full `fts` walk of every app's data directory when the filesystem has no quota support, so the redundant call is real work on exactly the devices least able to afford it.
- The target was derived from a **different reading** than the one the user's "freed" figure is computed from. Asking PMS to reclaim *X* and then telling the user *Y* was freed, where X and Y are two measurements of the same thing taken at different instants, is a gratuitous way to be inconsistent.

`cacheTrimTargetBytes` now takes the cache total as a parameter, and `measuringCacheFreed` hands its before-reading down to the clear lambda. The two uses are provably the same number. The per-app path ignores the argument.

### 2. Verified, left alone — the sweep does not clear `code_cache`, but the measurement counts it

`StorageStats.cacheBytes` includes `code_cache` on every device, by both accounting paths:

- Manual: `InstalldNativeService.cpp:2455` — `if (!strcmp(name, "cache") || !strcmp(name, "code_cache")) { stats->cacheSize += size; }`
- Quota: `collectQuotaStats` bills `stats->cacheSize` from the cache GID / cache project ID, and `prepare_app_cache_dir(path, "code_cache", 02771, uid, cacheGid, projectIdCache)` (line 787) applies that same GID to `code_cache`.

`PerUserCommands.clearCachePaths` names only `cache`. So the sweep leaves `code_cache` behind while the before/after subtraction still counts it.

**Not fixed here, deliberately.** `clearCachePaths` is shared with the per-app `clearCache`, and *there* the omission is correct: Settings' per-app "Clear cache" goes through `clearAppData(FLAG_CLEAR_CACHE_ONLY)`, which is a distinct flag from `FLAG_CLEAR_CODE_CACHE_ONLY` and does not touch `code_cache`. Matching Settings per-app was the right call and stays.

The global path is standing in for `freeCache`, which *does* take `code_cache` — so closing the gap means splitting the path list into per-app and global variants. That is a behaviour change beyond the reported bug, and on the primary path it is already masked: with a correct target, the trim runs first and PMS clears both. The residue is root-without-usage-access, where nothing is measured anyway. Filed rather than smuggled in.

### 3. Verified, closed — the target does not overshoot on a multi-user device

Worth checking, because the reporter tested HyperOS 3, where Second Space is a real second Android user.

`installd`'s `freeCache` builds its trackers from `get_known_users()` and walks `ce_path`, `de_path`, **and** `findDataMediaPath(uuid, userId) + "/Android/data/"` (`InstalldNativeService.cpp:2140-2146`) — every user, and external cache too. Thor's `totalCacheBytes()` measures Thor's user only. So the target is derived from a strict subset of what PMS can reclaim, which is the safe direction: PMS can always reach it without exhausting the cache rungs, and it stops the moment `getUsableSpace()` clears the bar.

The one consequence is that PMS may satisfy a Thor-user-sized target partly out of a work profile's or Second Space's cache, leaving some of Thor's user's cache behind. That is exactly what the sweep beneath it catches — and it is why the sweep staying scoped to `thorUserId` while the trim is volume-wide is not the asymmetry it looks like.

Also confirmed while checking this: `Environment.getDataDirectory().usableSpace` is `statvfs`'s `f_bavail`, the same field `File.getUsableSpace()` gives `system_server`. PMS is not root either, so both sides see the same number past the reserved blocks — which is the whole point of reading it from the same `File` PMS resolves.

File: app/src/main/java/com/valhalla/thor/data/gateway/ShizukuSystemGateway.kt [hunk 1]
```diff
@@ -69,15 +69,26 @@ class ShizukuSystemGateway(
...
             return Result.failure(
                 Exception(
-                    "Shizuku: could not read how much free space the volume has, and `pm trim-caches` " +
-                        "needs that number — passing a round one would let PackageManagerService walk " +
-                        "past the cache rungs into pruning apps."
+                    "Shizuku: clearing every app's cache needs Usage Access, because `pm trim-caches` " +
+                        "has to be told how many bytes to reclaim and only that permission can measure " +
+                        "the cache. Grant Thor usage access in Settings and try again."
                 )
             )
```
[the remark this answers is not in the document]
Comment (trinadhthatakula) [reply app/src/main/java/com/valhalla/thor/data/gateway/ShizukuSystemGateway.kt:86]: Fixed in `7c1c65a` — right, and the convention was already in this file: `ShizukuSystemGateway` takes a `Context` for exactly one reason, stated in its own constructor comment, which is that user-read refusals come out of resources.

Went one step wider than the flag. `DhizukuSystemGateway.clearAllCaches` had the same shape of message hardcoded, and localizing only the Shizuku one would have created a fresh inconsistency inside a single feature. What the two have in common is the test I used: each names an action the user can take — grant Usage Access, or switch privilege mode.

The diagnostics stay as literals. "`pm trim-caches` failed" gives the user nothing to do and matches its siblings across all three gateways; promoting those would be a separate, repo-wide change.

New strings: `clear_all_caches_requires_usage_access` and `clear_all_caches_unsupported_dhizuku`, across all five locales (en, ar, es, fr, zh-rCN) — `MissingTranslation` is fatal here, so a partial pass would not have built.
